### PR TITLE
Move hacking behavior to state classes

### PR DESCRIPTION
**Pull request by tgjones**
As a prerequisite to #879, I wanted to make `State` subclasses contain not just data, but also behavior. 

We've done this already with our [WeaponStateMachine](https://github.com/OpenSAGE/OpenSAGE/blob/master/src/OpenSage.Game/Logic/Object/Weapon/WeaponStates/WeaponStateMachine.cs), and it's pretty standard in any case. See [Super Simple State Machine](https://keithmaggio.wordpress.com/code/supersimplestatemachine/) for another example. 

Having `State` subclasses contain behavior and handle transitions to the next state should provide a nice framework for implementing more of these.

### CODE REVIEW

File: src/OpenSage.Game/Logic/Object/GameObject.cs [hunk 1]
```diff
@@ -1476,6 +1476,8 @@ public void Persist(StatePersister reader)
             reader.PersistUInt32(ref teamId);
             Team = GameContext.Game.TeamFactory.FindTeamById(teamId);
 
+            Owner = Team.Template.Owner;
```
Comment (tgjones) [inline src/OpenSage.Game/Logic/Object/GameObject.cs:1479]: This is unrelated to the PR. I just found that game objects are assigned to the wrong owner when loading a `.sav` file.

File: src/OpenSage.Game/Logic/AI/AIStates/StopHackingInternetState.cs [hunk 2]
```diff
@@ -4,13 +4,54 @@ namespace OpenSage.Logic.AI.AIStates
...
+        {
+            _gameObject.ModelConditionFlags.Set(ModelConditionFlag.Unpacking, false);
+            _gameObject.ModelConditionFlags.Set(ModelConditionFlag.FiringA, false);
+            _gameObject.ModelConditionFlags.Set(ModelConditionFlag.Packing, true);
+
+            _gameObject.GameContext.AudioSystem.PlayAudioEvent(_gameObject, _gameObject.Definition.UnitSpecificSounds.UnitPack?.Value);
+
+            var aiUpdate = (HackInternetAIUpdate)_gameObject.AIUpdate;
+
+            var frames = aiUpdate.GetVariableFrames(aiUpdate.ModuleData.PackTime);
```
Comment (charliefoxtwo) [inline src/OpenSage.Game/Logic/AI/AIStates/StopHackingInternetState.cs:28]: It would be nice if we could pass this in (or perhaps just expose a Set method or similar? really it would be nice if we could _construct_ this state with this parameter) so we don't have to access the AIUpdate object from here, and then `ModuleData` and `GetVariableFrames` could remain private. However, given the way we currently set states, that could be a more complex task.

The same goes for StartHackingInternetState as well. Perhaps some sort of `transitionTimer` or similar could be added and utilized for other modules as well?

**Review by charliefoxtwo — APPROVED** [review]
Nice! I like that we can keep the State data within their respective classes now.

File: src/OpenSage.Game/Logic/AI/AIStates/StopHackingInternetState.cs (hunk 2 above)
Comment (tgjones) [reply src/OpenSage.Game/Logic/AI/AIStates/StopHackingInternetState.cs:28]: > Perhaps some sort of transitionTimer or similar could be added and utilized for other modules as well?

That could be nice, but I think I'd prefer to implement a few more modules and then see what common behaviour we want to extract into helper classes. I just don't know yet what we'll need.
